File: 2022/day11.cpp
// https://adventofcode.com/2022/day/11

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <functional>
#include <limits>
#include <print>
#include <ranges>
#include <stdexcept>

using Item = u64;

enum class OperationType {
  add,
  multiply,
  square,
};

struct Operation {
  OperationType type = OperationType::add;
  Item value = 0;
};

struct Test {
  Item divisible_by = 1;
  usize on_true = 0;
  usize on_false = 0;
};

struct Monkey {
  Vec<Item> items;
  Operation operation;
  Test test;
};
using Monkeys = Vec<Monkey>;

fn after_split(str line, str delimiter) -> str {
  return aoc::split_once(line, delimiter)[1];
}

fn parse_operation(str line) -> Operation {
  let expression = after_split(line, " = ");
  let parts = aoc::split_to_array<3>(expression, ' ');
  let op = parts[1];
  let rhs = parts[2];
  if (op == "+") {
    return {OperationType::add, aoc::to_number<Item>(rhs)};
  }
  if (op == "*") {
    if (rhs == "old") {
      return {OperationType::square};
    }
    return {OperationType::multiply, aoc::to_number<Item>(rhs)};
  }
  throw std::runtime_error("Invalid operation " + String{op});
}

fn parse_monkey(str block) -> Monkey {
  let lines = aoc::split_to_vec(block, '\n');
  return Monkey{
      .items = aoc::split<Vec<Item>>(after_split(lines[1], ": "), ", "),
      .operation = parse_operation(lines[2]),
      .test =
          Test{
              .divisible_by =
                  aoc::to_number<Item>(after_split(lines[3], "by ")),
              .on_true =
                  aoc::to_number<usize>(after_split(lines[4], "monkey ")),
              .on_false =
                  aoc::to_number<usize>(after_split(lines[5], "monkey ")),
          },
  };
}

auto parse(String const& filename) -> Monkeys {
  return aoc::split(aoc::trim(aoc::read_file(filename)), "\n\n") |
         stdv::transform(parse_monkey) |
         aoc::ranges::to<Monkeys>();
}

fn inspect(Operation operation, Item old) -> Item {
  switch (operation.type) {
    case OperationType::add:
      return old + operation.value;
    case OperationType::multiply:
      return old * operation.value;
    case OperationType::square:
      return old * old;
  }
  AOC_UNREACHABLE("Invalid operation");
}

template <usize num_rounds, Item relief_factor>
fn solve_case(Monkeys const& input) -> Item {
  auto monkeys = input;
  auto supermodulo = std::numeric_limits<Item>::max();
  if constexpr (relief_factor == 1) {
    supermodulo = 1;
    for (let& monkey : monkeys) {
      supermodulo *= monkey.test.divisible_by;
    }
  }
  auto num_inspected = Vec<Item>(monkeys.size(), 0);

  for (let _ : Range{0uz, num_rounds}) {
    for (usize monkey_id = 0; monkey_id < monkeys.size(); ++monkey_id) {
      auto& monkey = monkeys[monkey_id];
      auto items = std::move(monkey.items);
      num_inspected[monkey_id] += items.size();

      for (let item : items) {
        let worry_level =
            inspect(monkey.operation, item % supermodulo) / relief_factor;
        let next_monkey = ((worry_level % monkey.test.divisible_by) == 0)
                              ? monkey.test.on_true
                              : monkey.test.on_false;
        monkeys[next_monkey].items.push_back(worry_level);
      }
    }
  }

  stdr::sort(num_inspected, std::greater{});
  return num_inspected[0] * num_inspected[1];
}

int main() {
  std::println("Part 1");
  let example = parse("day11.example");
  AOC_EXPECT_RESULT(10605, (solve_case<20, 3>(example)));
  let input = parse("day11.input");
  AOC_EXPECT_RESULT(98280, (solve_case<20, 3>(input)));

  std::println("Part 2");
  AOC_EXPECT_RESULT(2713310158, (solve_case<10000, 1>(example)));
  AOC_EXPECT_RESULT(17673687232, (solve_case<10000, 1>(input)));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day11.rs
use itertools::Itertools;

#[derive(Clone)]
enum Operation {
    Add(u64),
    Mul(u64),
    Square,
}

#[derive(Clone)]
struct Monkey {
    items: Vec<u64>,
    operation: Operation,
    test: (u64, usize, usize),
}

impl Monkey {
    fn inspect(&self, old: u64) -> u64 {
        match self.operation {
            Operation::Add(add) => old + add,
            Operation::Mul(mul) => old * mul,
            Operation::Square => old * old,
        }
    }
}

fn parse(filename: &str) -> Vec<Monkey> {
    let split_once = |it: &mut std::str::Lines<'_>, delimiter| {
        it.next()
            .unwrap()
            .split_once(delimiter)
            .unwrap()
            .1
            .to_string()
    };
    aoc::file::read_string(filename)
        .trim()
        .split("\n\n")
        .map(|monkey| {
            let lines = monkey.lines();
            let mut line_it = lines.into_iter();
            line_it.next();
            let items = split_once(&mut line_it, ": ");
            let operation = split_once(&mut line_it, " = ");
            let (_, op, rhs) = operation.split_whitespace().tuples().next().unwrap();
            let divisor = split_once(&mut line_it, "by ");
            let true_monkey = split_once(&mut line_it, "monkey ");
            let false_monkey = split_once(&mut line_it, "monkey ");
            Monkey {
                items: items.split(", ").map(|s| s.parse().unwrap()).collect(),
                operation: match op {
                    "+" => Operation::Add(rhs.parse().unwrap()),
                    "*" => {
                        if (rhs == "old") {
                            Operation::Square
                        } else {
                            Operation::Mul(rhs.parse().unwrap())
                        }
                    }
                    _ => unreachable!("Invalid operation"),
                },
                test: (
                    divisor.parse().unwrap(),
                    true_monkey.parse().unwrap(),
                    false_monkey.parse().unwrap(),
                ),
            }
        })
        .collect_vec()
}

fn solve_case<const NUM_ROUNDS: usize, const RELIEF: u64>(monkeys: &Vec<Monkey>) -> u64 {
    let supermodulo = if (RELIEF == 1) {
        monkeys.iter().map(|monkey| monkey.test.0).product::<u64>()
    } else {
        u64::MAX
    };
    let mut monkeys = monkeys.clone();
    let num_monkeys = monkeys.len();
    let mut num_inspected = vec![0; num_monkeys];
    for _ in 0..NUM_ROUNDS {
        for monkey_id in 0..num_monkeys {
            let items = monkeys[monkey_id].items.clone();
            for worry_level in items {
                let worry_level = monkeys[monkey_id].inspect(worry_level % supermodulo) / RELIEF;
                let next_index = if ((worry_level % monkeys[monkey_id].test.0) == 0) {
                    monkeys[monkey_id].test.1
                } else {
                    monkeys[monkey_id].test.2
                };
                monkeys[next_index].items.push(worry_level);
            }
            num_inspected[monkey_id] += monkeys[monkey_id].items.len();
            monkeys[monkey_id].items.clear();
        }
    }
    num_inspected.sort_unstable_by(|a, b| b.cmp(a));
    (num_inspected[0] as u64) * (num_inspected[1] as u64)
}

fn main() {
    println!("Part 1");
    let example = parse("day11.example");
    assert_eq!(10605, solve_case::<20, 3>(&example));
    let input = parse("day11.input");
    assert_eq!(98280, solve_case::<20, 3>(&input));

    println!("Part 2");
    assert_eq!(2713310158, solve_case::<10000, 1>(&example));
    assert_eq!(17673687232, solve_case::<10000, 1>(&input));
}
